Support edma3 for s32z270dc2 board

haduongquang opened the pull request:
This PR introduces NXP S32 EDMA3 driver for SoC NXP S32Z27 and enables its usage for board s32z270dc2

Tests:

tests\drivers\dma\chan_blen_transfer:
`SUITE PASS - 100.00% [dma_m2m]: pass = 4, fail = 0, skip = 0, total = 4 duration = 8.063 seconds`

tests\drivers\dma\chan_link_transfer:
`SUITE PASS - 100.00% [dma_m2m_link]: pass = 3, fail = 0, skip = 0, total = 3 duration = 6.047 seconds`

tests\drivers\dma\loop_transfer:
`SUITE PASS - 100.00% [dma_m2m_loop]: pass = 3, fail = 0, skip = 0, total = 3 duration = 1.103 seconds`

tests\drivers\dma\scatter_gather:
`SUITE PASS - 100.00% [dma_m2m_sg]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.076 seconds`

haduongquang committed:
boards/tests: dma: enable dma for S32Z270

Enable dma boards/test for S32Z270.

The non-cacheable memory used for DMA tests (chan_blen_transfer and
loop_transfer) is split from the system SRAM.

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

Dat-NguyenDuy commented on boards/nxp/s32z2xxdc2/doc/index.rst:
```diff
@@ -163,6 +165,16 @@ ADC is provided through ADC SAR controller with 2 instances. Each ADC SAR instan
+EDMA
+====
+
+The EDMA modules are categorized into one EDMA4 instance (instance 3) and four EDMA3 instances
```
The PR description said this supports eDMA3, why you mention eDMA4 here

haduongquang replied:
I removed eDMA4

Dat-NguyenDuy commented on tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:
```diff
@@ -0,0 +1,29 @@
+		sram2: memory@35870000 {
+			compatible = "zephyr,memory-region", "mmio-sram";
+			reg = <0x35870000 DT_SIZE_K(64)>;
+			zephyr,memory-region = "SRAM3";
```
`SRAM2`? or perhaps use `sram_nocache` for node lable and `SRAM_NOCACHE` for `zephyr,memory-region`. Use the same approach for all places

haduongquang replied:
I updated

Dat-NguyenDuy replied:
why not `sram2`?

haduongquang replied:
I updated `sram_nocache` for node lable and `SRAMNOCACHE` for `zephyr,memory-region`.

Dat-NguyenDuy commented on tests/drivers/dma/loop_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:
```diff
@@ -0,0 +1,29 @@
+		sram2: memory@35870000 {
+			compatible = "zephyr,memory-region", "mmio-sram";
+			reg = <0x35870000 DT_SIZE_K(64)>;
+			zephyr,memory-region = "SRAM3";
```
`SRAM2`?

Dat-NguyenDuy commented on tests/drivers/dma/chan_blen_transfer/boards/s32z2xxdc2_s32z270_rtu1.overlay:
```diff
@@ -0,0 +1,29 @@
+
+/ {
+	soc {
+		sram1: memory@35780000 {
```
You don't need to re-define the whole node here, just use something like this (but out of devicetree root `/{ }` )
```
&sram1 {
     reg = <0x31780000 DT_SIZE_K(960)>;
}
```
Do this for all places. Then please mention in the commit message that the non-cacheable memory used for DMA tests is split from the system SRAM

haduongquang replied:
Thanks, I updated.

Dat-NguyenDuy replied:
Where is this update `Then please mention in the commit message that the non-cacheable memory used for DMA tests is split from the system SRAM`? put something like that in commit msg or in code pls

haduongquang replied:
I updated the commit msg also.

haduongquang committed:
drivers: dma: support multiple instances

Support multiple instances for dma.

Memset struct data and tcdpool on initialization is not necessary.
Therefore remove it.

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang commented:
Added commit to Split `tcdpool` per device that can be used for multiple instances.

manuargue commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -716,17 +714,26 @@ static int dma_mcux_edma_init(const struct device *dev)
 	DMAMUX_BASE_INIT_DEFINE(n)						\
 	static void dma_imx_config_func_##n(const struct device *dev);		\
+	static __aligned(32) EDMA_TCDPOOL_CACHE_ATTR edma_tcd_t			\
+	tcdpool##n[DT_INST_PROP(n, dma_channels)][CONFIG_DMA_TCD_QUEUE_SIZE];	\
```
```suggestion
	dma_tcdpool_##n[DT_INST_PROP(n, dma_channels)][CONFIG_DMA_TCD_QUEUE_SIZE];	\
```

haduongquang replied:
I updated

manuargue commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -716,17 +714,26 @@ static int dma_mcux_edma_init(const struct device *dev)
 										\
-	struct dma_mcux_edma_data dma_data_##n;					\
+	static struct call_back data_callback_##n[DT_INST_PROP(n, dma_channels)];\
+	static ATOMIC_DEFINE(channels_atomic_##n, DT_INST_PROP(n, dma_channels));\
```
```suggestion
	static ATOMIC_DEFINE(dma_channels_atomic_##n, DT_INST_PROP(n, dma_channels));\
```

haduongquang replied:
I updated

manuargue commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -716,17 +714,26 @@ static int dma_mcux_edma_init(const struct device *dev)
 	};									\
 										\
-	struct dma_mcux_edma_data dma_data_##n;					\
+	static struct call_back data_callback_##n[DT_INST_PROP(n, dma_channels)];\
```
```suggestion
	static struct call_back dma_data_callback_##n[DT_INST_PROP(n, dma_channels)];\
```

haduongquang replied:
I updated

haduongquang commented:
@teburd @manuargue @Dat-NguyenDuy please help me take a look into this PR. All comments addressed. Thanks

haduongquang commented:
Hi @teburd, Could you please help me take a look into this PR?

haduongquang commented:
I rebased and fixed the conflict.

dleach02 commented:
@teburd I've add @manuargue as a co-assignee so we can move forward with this PR.

manuargue commented:
@haduongquang HAL pr merged, please rebase

manuargue reviewed (changes requested):
blocking until https://github.com/zephyrproject-rtos/zephyr/pull/78314 is merged and update the manifest

haduongquang committed:
soc: nxp: s32ze: add support eDMA3 for S32Z270

Enable support EDMA for S32Z270.
Add eDMA3 instance 0, 1, 4 and 5 for S32Z270 devices.

Signed-off-by: Ha Duong Quang <ha.duongquang@nxp.com>

haduongquang commented:
> blocking until #78314 is merged and update the manifest

I rebased and updated the manifest

manuargue reviewed (approved):
dismiss https://github.com/zephyrproject-rtos/zephyr/pull/78368#pullrequestreview-2376856145, my mistake, first we need to merge this pr
